pkg/planner/core/plan_cache.go

```diff
@@ -57,16 +57,24 @@ func SetParameterValuesIntoSCtx(sctx base.PlanContext, isNonPrep bool, markers [
 	vars := sctx.GetSessionVars()
 	vars.PlanCacheParams.Reset()
 	for i, usingParam := range params {
-		val, err := usingParam.Eval(sctx.GetExprCtx().GetEvalCtx(), chunk.Row{})
-		if err != nil {
-			return err
-		}
-		if isGetVarBinaryLiteral(sctx, usingParam) {
-			binVal, convErr := val.ToBytes()
-			if convErr != nil {
-				return convErr
+		var (
+			val types.Datum
+			err error
+		)
+		if c, ok := usingParam.(*expression.Constant); ok {
+			val = c.Value
```

Comment:
This seems dangerous since `Constant` might contain `DeferredExpr`. Why not always use `Eval`?

Reply:
The constant is caused by inlining the read-only getvar. I'll add an `intest.Assert`.